File: skills/python-code-reviewer/examples/good-service.py
"""
Minimal "good" counterpart of ``bad-service.py``.

Each fix is annotated with a tag that maps to the issue list in
``examples/review-output.md`` so bad → good is traceable 1:1.
"""

import json
from typing import Optional

import httpx
from sqlalchemy import text
from sqlalchemy.orm import Session


class RateLimiter:
    """Per-process LRU limiter. For multi-process / multi-host use Redis."""

    def __init__(self, max_hits: int, window_seconds: float) -> None:
        self.max_hits = max_hits
        self.window = window_seconds
        self._buckets: dict[int, list[float]] = {}

    def hit(self, key: int, now: float) -> bool:
        bucket = self._buckets.setdefault(key, [])
        # drop expired entries
        cutoff = now - self.window
        bucket[:] = [t for t in bucket if t >= cutoff]
        if len(bucket) >= self.max_hits:
            return False
        bucket.append(now)
        return True


class OrderService:
    def __init__(self, db: Session, pay_client: httpx.AsyncClient, limiter: RateLimiter) -> None:
        self.db = db
        self.pay = pay_client
        self.limiter = limiter

    async def pay(self, user_id: int, order_id: int, now: float) -> dict:
        # Fix 1 (module-level state): rate limit goes through a class
        # instance, not a module dict; per-process LRU with explicit
        # expiry. Multi-host deployments should swap in Redis.
        if not self.limiter.hit(user_id, now):
            return {"ok": False, "reason": "rate_limited"}

        # Fix 2 (SQL injection + ownership): bound parameters via
        # SQLAlchemy text() + named binds; the user_id predicate closes
        # the IDOR window.
        row = self.db.execute(
            text("SELECT id, user_id, status, pay_token, amount FROM orders WHERE id = :id AND user_id = :uid"),
            {"id": order_id, "uid": user_id},
        ).first()
        if row is None:
            return {"ok": False, "reason": "not_found"}
        # Fix 3 (idempotency): paid-state short-circuit precedes any side
        # effect so a redelivery is a no-op.
        if row.status == "PAID":
            return {"ok": True}

        # Fix 4 (status race): conditional UPDATE on old status, and
        # rollback on rowcount != 1 to release the row's lock.
        result = self.db.execute(
            text("UPDATE orders SET status = 'PAID' WHERE id = :id AND status = 'UNPAID'"),
            {"id": order_id},
        )
        if result.rowcount != 1:
            self.db.rollback()
            return {"ok": False, "reason": "concurrent_update"}

        # Fix 5 (async HTTP): httpx.AsyncClient is awaited; the call
        # carries the request context so a slow upstream does not block
        # the event loop.
        try:
            resp = await self.pay.post(
                "http://pay.local/charge",
                json={"token": row.pay_token, "amount": row.amount},
            )
            resp.raise_for_status()
        except httpx.HTTPError as e:
            # Fix 6 (rollback on charge failure): don't leave the row in
            # PAYING state if the charge call fails.
            self.db.rollback()
            return {"ok": False, "reason": "charge_failed", "error": str(e)}

        self.db.commit()
        return {"ok": True}


# Fix 7 (pickle.loads): refuse pickle entirely; the API contract becomes
# "JSON or msgpack" so an attacker cannot supply a malicious blob that
# executes on deserialise.
def load_blob(blob: bytes):
    raise NotImplementedError("pickle is forbidden; use JSON or msgpack")


# Fix 8 (bare except): narrow the catch to the specific error types
# the handler can plausibly raise; re-raise as the domain error with
# `from e` so the original cause is preserved in the traceback.
def process_event(event):
    try:
        return event.handler()
    except (KeyError, ValueError) as e:
        raise RuntimeError("event handler failed") from e


# Fix 9 (N+1): one IN (?, ?, ...) query with bound parameters; the
# placeholders are built from the id count, not user input.
def find_orders(session: Session, ids):
    if not ids:
        return []
    placeholders = ",".join(f":id{i}" for i in range(len(ids)))
    params = {f"id{i}": v for i, v in enumerate(ids)}
    return session.execute(
        text(f"SELECT id, status FROM orders WHERE id IN ({placeholders})"),
        params,
    ).all()

File: skills/node-code-reviewer/examples/good-service.ts
/**
 * Minimal "good" counterpart of `bad-service.ts`.
 * Each fix is annotated with a tag that maps to the issue list in
 * `examples/review-output.md` so bad → good is traceable 1:1.
 */

import * as fs from "node:fs/promises";
import { Prisma, PrismaClient } from "@prisma/client";


class RateLimiter {
  private buckets = new Map<number, number[]>();
  constructor(private max: number, private windowMs: number) {}

  hit(key: number, now: number): boolean {
    const arr = this.buckets.get(key) ?? [];
    const cutoff = now - this.windowMs;
    const next = arr.filter((t) => t >= cutoff);
    if (next.length >= this.max) {
      this.buckets.set(key, next);
      return false;
    }
    next.push(now);
    this.buckets.set(key, next);
    return true;
  }
}


const prisma = new PrismaClient();
const limiter = new RateLimiter(100, 60_000);


export class OrderService {
  // Fix 1 (sync IO): async file read so the event loop is not blocked
  // by fs.readFileSync; the call awaits the OS-level completion.
  static async loadConfig(): Promise<unknown> {
    const buf = await fs.readFile("/etc/order/config.json");
    return JSON.parse(buf.toString("utf8"));
  }

  static async pay(userId: number, orderId: number, now: number): Promise<{ ok: boolean }> {
    // Fix 2 (module-level state): rate limit goes through a class
    // instance, not a module-level Map; per-process LRU. For multi-host
    // deployments swap in Redis.
    if (!limiter.hit(userId, now)) return { ok: false, reason: "rate_limited" };

    // Fix 3 (filter allowlist): explicit where shape, no spread of a
    // dynamic object that could include arbitrary fields.
    const order = await prisma.order.findFirst({
      where: { id: orderId, userId },
    });
    if (!order) return { ok: false, reason: "not_found" };
    // Fix 4 (idempotency): the paid-state short-circuit precedes any
    // side effect so a redelivery is a no-op.
    if (order.status === "PAID") return { ok: true };

    // Fix 5 (status race + bound SQL): Prisma.sql template tag binds
    // the parameter; the WHERE clause guards on old status so two
    // concurrent pay calls cannot both flip the row.
    const result = await prisma.$executeRaw(
      Prisma.sql`UPDATE orders SET status = 'PAID' WHERE id = ${orderId} AND status = 'UNPAID'`,
    );
    if (result !== 1) {
      return { ok: false, reason: "concurrent_update" };
    }
    return { ok: true };
  }
}

// Fix 6 (prototype pollution): refuse __proto__ / constructor /
// prototype keys before they reach Object.assign, and use a
// null-prototype target as defence in depth.
const FORBIDDEN_KEYS = new Set(["__proto__", "constructor", "prototype"]);

export function mergeConfig(input: string): Record<string, unknown> {
  const parsed = JSON.parse(input) as Record<string, unknown>;
  for (const k of Object.keys(parsed)) {
    if (FORBIDDEN_KEYS.has(k)) {
      throw new Error(`forbidden config key: ${k}`);
    }
  }
  return Object.assign(Object.create(null), parsed);
}

// Fix 7 (lost await): await the work so the rejection surfaces as a
// real Promise rejection the caller can handle, not as an
// unhandledRejection event.
export async function backgroundJob(payload: string): Promise<void> {
  try {
    await doWork(payload);
  } catch (err) {
    // Re-throw so the caller / task supervisor sees the failure.
    throw err;
  }
}

async function doWork(_p: string): Promise<void> {
  throw new Error("simulated failure");
}
